SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp

```diff
@@ -242,51 +252,137 @@ void AEventTracingTest::GatherDataFromFile(const FString& FilePath)
 	Stream = nullptr;
 }
 
-bool AEventTracingTest::CheckEventTraceCause(const FString& SpanIdString, const TArray<FName>& CauseEventNames,
-											 int MinimumCauses /*= 1*/) const
+int32 AEventTracingTest::GetTraceSpanCount(const TraceSource Source) const
+{
+	const TraceItemsData* SourceTraceItems = TraceItems.Find(Source);
+	return SourceTraceItems != nullptr ? SourceTraceItems->Spans.Num() : 0;
+}
+
+int32 AEventTracingTest::GetTraceEventCount(const TraceSource Source) const
 {
-	const TArray<FString>* Causes = TraceSpans.Find(SpanIdString);
-	if (Causes == nullptr || Causes->Num() < MinimumCauses)
+	const TraceItemsData* SourceTraceItems = TraceItems.Find(Source);
+	if (SourceTraceItems == nullptr)
 	{
+		return 0;
+	}
+
+	int32 TotalCount = 0;
+	for (const auto& Pair : SourceTraceItems->EventCount)
+	{
+		TotalCount += Pair.Value;
+	}
+	return TotalCount;
+}
+
+int32 AEventTracingTest::GetTraceEventCount(const TraceSource Source, const FName TraceEventType) const
+{
+	const TraceItemsData* SourceTraceItems = TraceItems.Find(Source);
+	if (SourceTraceItems == nullptr)
+	{
+		return 0;
+	}
+
+	const int32* Count = SourceTraceItems->EventCount.Find(TraceEventType);
+	return Count != nullptr ? *Count : 0;
+}
+
+FString AEventTracingTest::FindRootSpanId(const FString& SpanId) const
+{
+	FString Cause = SpanId;
+	ForEachTraceSource([&SpanId, &Cause](const TraceItemsData& SourceTraceItems) {
+		const TArray<FString>* Causes = SourceTraceItems.Spans.Find(SpanId);
+		if (Causes != nullptr && Causes->Num() > 0)
+		{
+			Cause = (*Causes)[0];
+			return true;
+		}
 		return false;
+	});
+
+	if (Cause != SpanId)
+	{
+		const FString NextCause = FindRootSpanId(Cause);
+		return NextCause == Cause ? Cause : NextCause;
 	}
+	return SpanId;
+}
 
-	for (const FString& CauseSpanIdString : *Causes)
+void AEventTracingTest::ForEachTraceSource(TFunctionRef<bool(const TraceItemsData& SourceTraceItems)> Predicate) const
+{
+	for (int32 i = 0; i < TraceSource::Count; ++i)
 	{
-		const FName* CauseEventName = TraceEvents.Find(CauseSpanIdString);
-		if (CauseEventName == nullptr)
+		TraceSource Source = static_cast<TraceSource>(i);
+		const TraceItemsData* SourceTraceItems = TraceItems.Find(Source);
+		if (SourceTraceItems != nullptr)
 		{
-			return false;
+			if (Predicate(*SourceTraceItems))
+			{
+				break;
+			}
 		}
-		if (!CauseEventNames.Contains(*CauseEventName))
+	}
+}
+
+bool AEventTracingTest::CheckEventTraceCause(const FString& SpanIdString, const TArray<FName>& CauseEventNames,
+											 int MinimumCauses /*= 1*/) const
+{
+	bool bSuccess = true;
+	ForEachTraceSource([this, &SpanIdString, &MinimumCauses, &CauseEventNames, &bSuccess](const TraceItemsData& SourceTraceItems) {
```

Comment:
CheckEventTraceCause functionally is still pretty much the same. It now requires ForEachTraceSource to be used as the data it previously operated on was held inside one container. Now the data is spread over different map entries.